### PR TITLE
bpf: don't use fixed size integer types from stdint.h

### LINKED ISSUE #368: `fatal error: 'gnu/stubs-32.h' file not found` while building
If you are seeing the following or a similar error while building Cilium
or while building programs for endpoints:
```
    In file included from /usr/include/features.h:392:
    /usr/include/gnu/stubs.h:7:11: fatal error: 'gnu/stubs-32.h' file not found
    # include <gnu/stubs-32.h>
              ^
    1 error generated.
```
Then your system does not have the 32-bit versions of Glibc installed. Install
them via your distribution's package, e.g. for Fedora:
```
    dnf install glibc-devel.i686
```

### LINKED ISSUE #8529: trial build cilium 1.5.3 for ppc64le failed as missing "__powerpc64__" defined
## Support
trial build cilium 1.5.3 for ppc64le failed as missing ```__powerpc64__``` defined

I just tried in OBS tool to build Cilium 1.5.3 on openSUSE for ppc64le and it failed as reported at (1)
as far as I understand the failure in stub-32.h
is because __WORDSIZE is not set to 64 in stubs.h
itself because ```__powerpc64__``` not set in bits/wordsize.h

I am wondering what should be added in  contrib/packaging/rpm/cilium.spec.envsubst or bpf/Makefile.bpf

(1) https://build.opensuse.org/package/live_build_log/home:michel_mno:branches:devel:kubic/cilium/openSUSE_Factory_PowerPC/ppc64le
with following copy/past lines:

```
[  286s] make[1]: Entering directory '/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf'
[  286s] set -e; \
[  286s]  :  " [-DSKIP_DEBUG]"; clang -DSKIP_DEBUG -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DHAVE_LPM_MAP_TYPE]"; clang -DHAVE_LPM_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DHAVE_LRU_MAP_TYPE]"; clang -DHAVE_LRU_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV4]"; clang -DENABLE_IPV4 -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE]"; clang -DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]"; clang -DENABLE_IPV4 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV6]"; clang -DENABLE_IPV6 -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE]"; clang -DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]"; clang -DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV6 -DENABLE_IPV4]"; clang -DENABLE_IPV6 -DENABLE_IPV4 -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_IPV4 -DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE]"; clang -DENABLE_IPV4 -DENABLE_IPV6 -DHAVE_LPM_MAP_TYPE -DHAVE_LRU_MAP_TYPE -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_HOST_REDIRECT -DENABLE_IPV4 -DENABLE_IPV6]"; clang -DENABLE_HOST_REDIRECT -DENABLE_IPV4 -DENABLE_IPV6 -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll;   :  " [-DENABLE_HOST_REDIRECT -DENABLE_IPV4 -DENABLE_IPV6 -DENABLE_NAT46]"; clang -DENABLE_HOST_REDIRECT -DENABLE_IPV4 -DENABLE_IPV6 -DENABLE_NAT46 -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include -I/home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf -D__NR_CPUS__=4 -O2 -g -target bpf -emit-llvm -Wall -Werror -Wno-address-of-packed-member -Wno-unknown-warning-option -c bpf_lxc.c -o bpf_lxc.ll; llc -march=bpf -mcpu=probe -mattr=dwarfris -o /dev/null bpf_lxc.ll; 
[  286s] In file included from bpf_lxc.c:18:
[  286s] In file included from /home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/node_config.h:29:
[  286s] In file included from /home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/lib/utils.h:21:
[  286s] In file included from /home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include/bpf/api.h:11:
[  286s] In file included from /home/abuild/rpmbuild/BUILD/go/src/github.com/cilium/cilium/bpf/include/linux/type_mapper.h:4:
[  286s] In file included from /usr/lib64/clang/8.0.0/include/stdint.h:63:
[  286s] In file included from /usr/include/stdint.h:26:
[  286s] In file included from /usr/include/bits/libc-header-start.h:33:
[  286s] In file included from /usr/include/features.h:474:
[  286s] /usr/include/gnu/stubs.h:8:11: fatal error: 'gnu/stubs-32.h' file not found
[  286s] # include <gnu/stubs-32.h>
[  286s]           ^~~~~~~~~~~~~~~~
[  286s] 1 error generated.
```

### PR DESCRIPTION
**Pull request by tklauser**
Use stddef.h to get size_t, use kernel definitions for fixed size types
where appropriate (e.g. uint32_t -> __u32)

This allows to get rid of the x32 libc header dependecy to pull in the
GNU libc stub headers. It should thus more easily allow to compile the
BPF programs on platforms other than amd64 (e.g. arm64, ppc64).

Also clean up some other unused includes.

For some reason these changes lead to clang complaining about several
functions defined in .h files being unused. Mark these as __maybe_unused
or if they are used in a single place move them to the .c file.

Fixes #368
Fixes #8529

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

  ```release-note
BPF programs no longer depend on libc headers.
  ```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10204)

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Review by vadorovsky — COMMENTED** [review]
If the goal of that pull request is only getting rid of `libc6-dev-i386` dependency, then I think there is an easier way to do that - adding `-D__x86_64__` to CFLAGS when compiling BPF programs. After using that flag, including `gnu/stubs.h` will result in including only `gnu/stubs-64.h` and `gnu/stubs-32.h` will be not included anymore. That's because the content of `gnu/stubs.h` usually looks like that (at least on Ubuntu and openSUSE):

```
#if !defined __x86_64__
# include <gnu/stubs-32.h>
#endif
#if defined __x86_64__ && defined __LP64__
# include <gnu/stubs-64.h>
#endif
#if defined __x86_64__ && defined __ILP32__
# include <gnu/stubs-x32.h>
#endif
```

If you want, I can submit a PR to try that approach.

**Comment by mrostecki** [comment]
Because, as far as I understand your PR correctly, you are not getting rid of dependency on glibc in general?

**Comment by tklauser** [comment]
> Because, as far as I understand your PR correctly, you are not getting rid of dependency on glibc in general?

The idea of this PR is to no longer depend on any glibc headers for BPF programs. Getting rid of `libc-dev-i386` is just a nice side-effect.

It's still work-in-progress (see the CI failures), but I think I'm almost there.

**Comment by mrostecki** [comment]
OK, that's great, then LGTM.

**Comment by borkmann** [comment]
> The idea of this PR is to no longer depend on any glibc headers for BPF programs. Getting rid of `libc-dev-i386` is just a nice side-effect.

Yep, this is the reason why we need this PR here, which is also a big step forward for multi-arch support.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

File: pkg/datapath/linux/requirements.go [hunk 1]
```diff
@@ -183,16 +183,6 @@ func CheckMinRequirements() {
 				log.Warn("llc version was compiled in debug mode, expect higher latency!")
 			}
 		}
-		// /usr/include/gnu/stubs-32.h is installed by 'glibc-devel.i686' in fedora
```
Comment (Jianlin-lv) [inline pkg/datapath/linux/requirements.go:186]: As far as I know，this section is for people that are not running cilium in a container; 
I have a question about whether need to delete this section code?

Comment (vadorovsky) [reply pkg/datapath/linux/requirements.go:186]: After this PR, Cilium does not depend on glibc header at all, so yes, that code can be removed.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
CI finally passed on the latest iteration. PTAL.

**Review by borkmann — APPROVED** [review]

File: bpf/bpf_alignchecker.c [hunk 2]
```diff
@@ -27,6 +25,7 @@
 #include "lib/conntrack.h"
 #include "lib/dbg.h"
 #include "lib/drop.h"
+#define SKIP_UNDEF_LPM_LOOKUP_FN
```
Comment (joestringer) [inline bpf/bpf_alignchecker.c:28]: Was this related, or just another cleanup that got caught up in this PR?

File: bpf/include/linux/byteorder.h [hunk 3]
```diff
@@ -1,9 +1,7 @@
 #ifndef _ASM_X86_BYTEORDER_H
 #define _ASM_X86_BYTEORDER_H
 
-#include <endian.h>
-
-#if __BYTE_ORDER == __LITTLE_ENDIAN
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Comment (joestringer) [inline bpf/include/linux/byteorder.h:4]: Do we need this header file at all any more?

File: test/helpers/cons.go [hunk 4]
```diff
@@ -257,6 +258,7 @@ var badLogMessages = map[string][]string{
 	NACKreceived:      nil,
 	RunInitFailed:     {"signal: terminated", "signal: killed"},
 	sizeMismatch:      nil,
+	emptyBPFInitArg:   nil,
```
Comment (joestringer) [inline test/helpers/cons.go]: I think these changes snuck in from another PR?

File: pkg/datapath/loader/compile.go [hunk 5]
```diff
@@ -192,7 +194,26 @@ func compileAndLink(ctx context.Context, prog *progInfo, dir *directoryInfo, deb
...
+)
+
+// getMachineName returns the machine hardware name of this host.
+func getMachineName() string {
+	unameOnce.Do(func() {
+		var uts unix.Utsname
+		err := unix.Uname(&uts)
+		if err == nil {
+			machineName = string(uts.Machine[:bytes.IndexByte(uts.Machine[:], 0)])
+		}
```
Comment (joestringer) [inline pkg/datapath/loader/compile.go:214]: Please at least log at debug level (or higher) on a failure here, with the error. Ignoring an error and falling back to x86-64 here means that this code could be completely broken and we'd never notice.

File: bpf/lib/utils.h [hunk 6]
```diff
@@ -88,18 +88,20 @@ static inline __u32 bpf_ktime_get_sec(void)
 	return (__u64)(bpf_ktime_get_nsec() / NSEC_PER_SEC);
 }
 
-#if __BYTE_ORDER == __LITTLE_ENDIAN
+#if defined(__BYTE_ORDER__) && defined(__ORDER_LITTLE_ENDIAN__) && \
+	__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Comment (joestringer) [inline bpf/lib/utils.h:92]: I don't know the strict rules on whose job it is to define these depending on your compiler, but given that we only ever use LLVM for this code and this is pretty similar to the existing code, I'm assuming this is correct and works.

**Review by joestringer — COMMENTED** [review]
Mostly nits below.

I'm not sure whether we need the `byteorder.h`  and endian header friends any more with this patch?

File: bpf/lib/utils.h (hunk 6 above)
Comment (borkmann) [reply bpf/lib/utils.h:92]: We explicitly need the byte order define from LLVM here. See also: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/lib/bpf/bpf_endian.h

File: bpf/bpf_alignchecker.c (hunk 2 above)
Comment (tklauser) [reply bpf/bpf_alignchecker.c:28]: Just another cleanup which was surfaced by several functions suddenly being complained about by the compiler as being unused after this change. I'm still not sure why there weren't any `-Wunused-function` errors before this change though.

File: bpf/include/linux/byteorder.h (hunk 3 above)
Comment (tklauser) [reply bpf/include/linux/byteorder.h:4]: Doesn't seem to be needed anymore. Will remove it.

File: bpf/lib/utils.h (hunk 6 above)
Comment (tklauser) [reply bpf/lib/utils.h:92]: These are defined by both GCC (https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html) and LLVM and it's e.g. what the Abseil C++ library uses: https://github.com/abseil/abseil-cpp/blob/c44657f55692eddf5504156645d1f4ec7b3acabd/absl/base/config.h#L444-L465

File: pkg/datapath/loader/compile.go (hunk 5 above)
Comment (tklauser) [reply pkg/datapath/loader/compile.go:214]: Good point. I think this might even deserve a warning?

File: test/helpers/cons.go (hunk 4 above)
Comment (tklauser) [reply test/helpers/cons.go]: Oops, indeed. This should have gone into #10230. Will remove.

File: bpf/include/linux/byteorder.h (hunk 3 above)
Comment (tklauser) [reply bpf/include/linux/byteorder.h:4]: Was a bit too quick to reply. Looks like it is still needed for `__constant_htonl` in `bpf/lib/utils.h`.

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Pull request merged**